clients/js/src/compat.ts

```diff
@@ -590,61 +594,83 @@ export async function fetchRuntimePublicKeyByChainId(
   chainId: number,
   opts?: { fetch?: typeof fetch },
 ): Promise<Uint8Array> {
-  const { defaultGateway: gatewayUrl } = NETWORKS[chainId];
-  if (!gatewayUrl)
+  const { defaultGateway } = NETWORKS[chainId];
+  if (!defaultGateway)
     throw new Error(
       `Unable to fetch runtime public key for network with unknown ID: ${chainId}.`,
     );
   const fetchImpl = opts?.fetch ?? globalThis?.fetch;
   const res = await (fetchImpl
-    ? fetchRuntimePublicKeyBrowser(gatewayUrl, fetchImpl)
-    : fetchRuntimePublicKeyNode(gatewayUrl));
+    ? fetchRuntimePublicKeyBrowser(defaultGateway, fetchImpl)
+    : fetchRuntimePublicKeyNode(defaultGateway));
   return getBytes(res.result.key);
 }
 
+function fromQuantity(x: number | string): number {
+  if (typeof x == 'string') {
+    if (x.startsWith('0x')) {
+      return parseInt(x, 16);
+    }
+    return parseInt(x); // Assumed to be base 10
+  }
+  return x;
+}
+
 /**
  * Picks the most user-trusted runtime calldata public key source based on what
  * connections are available.
+ *
+ * NOTE: MetaMask does not support Web3 methods it doesn't know about, so we have to
+ *       fall-back to manually querying the default gateway.
  */
 export async function fetchRuntimePublicKey(
   upstream: UpstreamProvider,
 ): Promise<Uint8Array> {
   const provider = 'provider' in upstream ? upstream['provider'] : upstream;
+  let chainId: number | undefined;
   if (provider) {
     let resp: any;
     // It's probably an EIP-1193 provider
     if ('request' in provider) {
+      const source = provider as EIP1193Provider;
       try {
-        const source = provider as EIP1193Provider;
         resp = await source.request({
           method: OASIS_CALL_DATA_PUBLIC_KEY,
           params: [],
         });
       } catch (ex) {
         // don't do anything, move on to try next
+        chainId = fromQuantity(
+          (await source.request({ method: 'eth_chainId' })) as string | number,
+        );
       }
     }
     // If it's a `send` provider
     else if ('send' in provider) {
+      const source = provider as {
+        send: (method: string, params: any[]) => Promise<any>;
+      };
       try {
-        const source = provider as {
-          send: (method: string, params: any[]) => Promise<any>;
-        };
         resp = await source.send(OASIS_CALL_DATA_PUBLIC_KEY, []);
       } catch (ex) {
         // don't do anything, move on to try chainId fetch
+        chainId = fromQuantity(await source.send('eth_chainId', []));
       }
     }
+    // Otherwise, we have no idea what to do with this provider!
+    else {
+      throw new Error(
+        'fetchRuntimePublicKey not sure how to handle non-request non-send provider!',
```

Comment:
Dunno if we want to be more firm 🤔 

```suggestion
        'fetchRuntimePublicKey does not support non-request non-send provider!',
```